Update upgrade.md with new preparation guide.

ellamaolson opened the pull request:
Add an outline for upgrade file and preparation section. Add three sections: (1) file structure, (2) language updates, (3) architectural changes to "Preparation" section. Add table comparing AngularJS and Angular versions of directives and components. Maintain this file as the ngUpgrade page.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
upgrade.md is out-of-date and needs a step-by-step preparation guide.
Issue Number: N/A


## What is the new behavior?
Step-by-step preparation guide added.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -1164,8 +1168,7 @@ in this project, and that's a bit special since it's an `ngResource`
 
-### Installing Angular
-
+## Installing Angular
```
Please keep the empty line.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -1,77 +1,73 @@
+ - [Upgrading with ngUpgrade](#upgrade)
+ - [Using Ahead-of-time compilation with hybrid apps](#aot)
+ - [PhoneCat Upgrade Tutorial](#phonecat)
+ - [Upgrading PhoneCat Tests](#tests)
```
Why keep this "table of contents" here? We already have a ToC on each guide.

ellamaolson replied:
Good point, I originally could not find the table of contents, and so created this one temporarily.

gkalpak replied:
The ToC is created at runtime based on the h1, h2, h3 elements on the guide (unless explicitly disabled with a `no-toc` class on the page's h1; e.g. on marketing pages).

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -1,77 +1,73 @@
+ - [PhoneCat Upgrade Tutorial](#phonecat)
+ - [Upgrading PhoneCat Tests](#tests)
+
+![terms](https://user-images.githubusercontent.com/27384475/41929674-b62c7292-792d-11e8-84ba-987c690ffb2d.png)
```
Images should be checked into the repo. E.g. in [aio/content/images/guide/upgrade][1].

[1]: https://github.com/ellamaolson/angular/tree/7b41344cd8ce3f2027b2eb902d395ae8aa4d3d46/aio/content/images/guide/upgrade

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -1,77 +1,73 @@
 to upgrade even large and complex applications without disrupting other
 business, because the work can be done collaboratively and spread over
-a period of time. The `upgrade` module in Angular has been designed to
-make incremental upgrading seamless.
```
Are you implying the upgrading process is not seamless? 🙊 😛

ellamaolson replied:
Not sure, just trying to reduce text.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -1,77 +1,73 @@
-much more easy to work with than others. There are a few key techniques
-and patterns that you can apply to future proof apps even before you
-begin the migration.
+# <a name="preparation"></a> Preparation for ngUpgrade
```
It is better to use `{@a some-hash}`. E.g.:

```
{@a preparation}
## Preparation for ngUpgrade
```

See also https://angular.io/guide/docs-style-guide#anchors.

gkalpak replied:
Each document must have only one h1 element (which is what `#` is converted to).

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -829,7 +869,7 @@ After this, the service is injectable anywhere in AngularJS code:
 </code-example>
 
-## Using Ahead-of-time compilation with hybrid apps
+## <a name="aot"></a> Using Ahead-of-time compilation with hybrid apps
```
Please use `{@a ...}` here (and below) too.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -1165,7 +1205,6 @@ factory. So you won't be doing anything to it in the preparation stage.
 You'll instead turn it directly into an Angular service.
 
 ### Installing Angular
-
```
Please keep empty lines around headings.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -53,25 +66,40 @@ aligned with Angular*.
-  and [Modularity](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#modularity)
-  rules define similar principles on a higher level of abstraction: Different parts of the
-  application should reside in different directories and NgModules.
+{@a rule} 
```
I think the anchors should be more descriptive. E.g. `#rule-of-1` vs `#rule` or `#module-loader` vs `#module`.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -53,25 +66,40 @@ aligned with Angular*.
 When an application is laid out feature per feature in this way, it can also be
 migrated one feature at a time. For applications that don't already look like
-this, applying the rules in the AngularJS style guide is a highly recommended
+this, applying the rules in the [AngularJS style guide](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md) is a highly recommended
```
Not sure this is necessary, since we are linking to the styleguide a few lines above.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -53,25 +66,40 @@ aligned with Angular*.
+
+{@a feature} 
+
+#### Feature-By-Feature
```
Shouldn't this be `Folders-by-Feature`?

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -53,25 +66,40 @@ aligned with Angular*.
 
+{@a lang} 
+
+### Language Updates
```
I don't think "Language" is a good title. Only "Migrating to TypeScript" is "language"-related. Maybe "Build Process" would be more accurate. (And "Migrating to TypeScript" might also fall under "Architecture" changes - not sure 🤔)

ellamaolson replied:
Wasn't sure either, will make the changes soon. :)

ellamaolson committed:
docs(aio): clarify and update upgrading preparation guide

Add an outline for upgrade file and preparation section. Add three sections: (1) file structure, (2) build process, (3) architectural changes to "Preparation" section. Add table comparing AngularJS and Angular versions of directives and components. Correct anchor tags. Maintain this file as the ngUpgrade page.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +153,42 @@ Additional steps like the following can give us much more bang for the buck:
+{@a updating-angularjs}
+ 
+####  Updating AngularJS
+There is a large shift in the architecture of AngularJS between versions 1.4 and 1.5. In order to prepare for ngUpgrade, you need to be using AngularJS versions 1.5+ and follow the guidelines for directive attributes. 
```
This sentence sounds confusing. I am not sure what architecture shift you are referring to nor what are the "guidelines for directive attributes" 😕

The main change in AngularJS 1.5 was the introduction of the `.component()` helper, which reduces the required boilerplate for creating "component directives" (which are directives that have certain properties). But `.component()` does not enable any new behavior that was not possible before. I.e., it is possible to create "component directives" with the old `.directive()` helper too.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +153,42 @@ Additional steps like the following can give us much more bang for the buck:
-In Angular, components are the main primitive from which user interfaces
-are built. You define the different portions of the UI as components and
-compose them into a full user experience.
+    component(name, options);
```
This line doesn't seem to belong here.

ellamaolson replied:
I was intending to show what a component directive looks like.

gkalpak replied:
If anything, this should be made a code block (e.g. by enclosing in tripple backticks:

> ```ts
> .component(name, options)
> ```

But I think there should be some ttext explaining what the code block shows.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +153,42 @@ Additional steps like the following can give us much more bang for the buck:
+|  AngularJS v1.4-   |  AngularJS v1.5   |  Angular v2.0+    |
+|:-----------------:|:-----------------:|:-----------------:|
+|  Directive        |Component Directive|      Component    |
+| are markers on a DOM element (i.e. attribute, element name, comment or CSS class) that tell AngularJS's HTML compiler to manipulate the DOM element and its children.  | are directives that define their own templates, controllers, and input/output bindings - the same things that Angular components define. | the main primitive from which user interfaces are built. You define the different portion ...
```
Nit: Comments are not "markers on a DOM element" (since comments are separate entities).

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +153,42 @@ Additional steps like the following can give us much more bang for the buck:
-  to the controller instead of using the `$scope`.
+Here is a comparison to show the transitions happening between AngularJS directives to Angular components.
+
+|  AngularJS v1.4-   |  AngularJS v1.5   |  Angular v2.0+    |
```
v1.5 --> v1.5+

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +153,42 @@ Additional steps like the following can give us much more bang for the buck:
+| are markers on a DOM element (i.e. attribute, element name, comment or CSS class) that tell AngularJS's HTML compiler to manipulate the DOM element and its children.  | are directives that define their own templates, controllers, and input/output bindings - the same things that Angular components define. | the main primitive from which user interfaces are built. You define the different portion ...
+
+##### Do's and Don'ts 
+Within component directives, there are some Do's and Don't you should follow to be Angular compatible. All of the following
```
Don't --> Don'ts (but it looks kind of weird)

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +153,42 @@ Additional steps like the following can give us much more bang for the buck:
+
+##### Do's and Don'ts 
+Within component directives, there are some Do's and Don't you should follow to be Angular compatible. All of the following
+guidelines fall within the definition of component directives. Redefine your 
```
Your what? :smiley:

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +153,42 @@ Additional steps like the following can give us much more bang for the buck:
+guidelines fall within the definition of component directives. Redefine your 
+
+To be Angular compatible, an AngularJS component directive **should** configure these attributes as follows:
+* `restrict: 'E'`. This is the default of restrict and should be the only way of writing the attribute. Components are always restricted to elements (i.e. restrict: 'E').
```
This is a prerequisite for a directive to be a "component directive" and is the only option when using `.component()`.
Also, it is not the default :confused:

ellamaolson replied:
Oh weird, I was told it was the default.

gkalpak replied:
To add some more details:
- The default for AngularJS general directives is `AE` (which means match both attributes or element names).
- For a directive to be a component directive it has to be `E`. (There are some corner cases where `A` might make sense, but are outside the scope of this comment :grin:)
- The `.component()` helper create a directive with `E` under the hood and does not support using a different value. (If you want to use a different value, you have to use the `.directive()` helper, not the `.component()` helper.)

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +153,42 @@ Additional steps like the following can give us much more bang for the buck:
+Within component directives, there are some Do's and Don't you should follow to be Angular compatible. All of the following
+guidelines fall within the definition of component directives. Redefine your 
+
+To be Angular compatible, an AngularJS component directive **should** configure these attributes as follows:
```
The following attributes are prerequisites for a directive to be considered a "component directive". Being Angular compatible is irrelevant here.

gkalpak commented:
@Jeremydavisvt, your points may be valid but out-of-scope for this PR. Let's track needed improvements in #24867.

IgorMinar commented on aio/content/guide/upgrade.md:
```diff
@@ -1711,7 +1758,9 @@ When you're done, this is what it should look like:
 it's time to say goodbye.
 
-## Appendix: Upgrading PhoneCat Tests
+{@a aot}
```
duplicate aot anchor?

IgorMinar commented on aio/content/guide/upgrade.md:
```diff
@@ -91,12 +115,21 @@ in the correct order.
+{@a package-manager} 
+
+#### Package Manager
+Make sure you are using either npm or yarn as your package manager instead of Bower, Babel, or anything else. 
```
Babel is not a package manager

IgorMinar reviewed (changes requested):
this PR seems currently incomplete

gkalpak commented on aio/content/guide/upgrade.md:
````diff
@@ -120,31 +157,43 @@ Additional steps like the following can give us much more bang for the buck:
-component directives are much easier to migrate to Angular than applications
-built with lower-level features like `ng-controller`,  `ng-include`, and scope
-inheritance.
+```
````
Use ```` ```js```` or ```` ```ts```` for better highlighting.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +157,43 @@ Additional steps like the following can give us much more bang for the buck:
+| are markers on a DOM element (i.e. attribute, element name, or CSS class) that tell AngularJS's HTML compiler to manipulate the DOM element and its children.  | are directives that define their own templates, controllers, and input/output bindings - the same things that Angular components define. | the main primitive from which user interfaces are built. You define the different portions of the ...
+
+#### Do's and Don'ts 
+Within component directives, there are some Do's and Don't you should follow to be be able to create a component directive. All of the following guidelines fall within the definition of component directives.
```
- _Don't_ --> _Don'ts_ (?)
- _to be be able_ --> _to be able_

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +157,43 @@ Additional steps like the following can give us much more bang for the buck:
+
+The following attributes are prerequisites that **should** be followed for a directive to be considered a component directive:
+
+* `restrict: 'E'`. This is a prerequisite for a directive to be a component directive and is the only option when using .component(). 
```
- Repeating that this is a prerequisite is redundant (since it is mentioned in the sentence right above).
- Wrap `.component()` in backticks (here and below), since it is code.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -1,37 +1,49 @@
 business, because the work can be done collaboratively and spread over
 a period of time. The `upgrade` module in Angular has been designed to
-make incremental upgrading seamless.
+-make incremental upgrading seamless. 
```
Bogus change.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -91,12 +115,25 @@ in the correct order.
+{@a using-a-package-manager} 
+
+#### Using a Package Manager
+A package manager eases finding, installing, maintaining, and uninstalling software for your application. It will maintain dependencies to ensure that all requirements for that package are installed and maintained. 
```
I am not sure what the "maintaining"/"maintained" part refers to.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -91,12 +115,25 @@ in the correct order.
-If part of the Angular upgrade plan is to also take TypeScript into use, it makes
+Make sure you are using either [npm](https://www.npmjs.com/) or [Yarn](https://yarnpkg.com/en/) as your package manager instead of Bower or anything else. 
+
+When you initialize a project with either package manager, a package.json file will be created that tells the package manager information about your package. Fields you see in a package.json file are **name, version, description, main, repository, author, license.** You can have additional fields like scripts, files, bin, dependencies, devDependencies, and many more. 
```
The "tells...information" part sounds weird (although I am not a antive speaker).
Maybe "...that contains infromation about your package"?

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +157,43 @@ Additional steps like the following can give us much more bang for the buck:
+
+#### Using Component Directives
+
+In AngularJS version 1.5, the [component directive](https://docs.angularjs.org/guide/component) was introduced to bridge the gap between AngularJS [directives](https://docs.angularjs.org/guide/directive) and Angular version 2.0+ [components](https://angular.io/guide/displaying-data). 
```
1. "The component directive was introduced" sounds inaccurate.
    I would either say "the component helper" or "the concept of component directives".

2. Maybe a link to https://angular.io/guide/architecture-components would be more appropriate for Angular 2+ components.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +157,43 @@ Additional steps like the following can give us much more bang for the buck:
-In Angular, components are the main primitive from which user interfaces
-are built. You define the different portions of the UI as components and
-compose them into a full user experience.
+**.component()** helper was introduced to reduce the required boilerplate for creating component directives (which are directives that have certain properties). It takes two arguments, the name of the component (as a string) and the component config object (Note that, unlike the .directive() method, this method does not take a factory function). 
```
- _Note_ --> _note_ (or put the dot after "config object")
- Wrap `.directive()` in backticks, since it is code.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -120,31 +157,43 @@ Additional steps like the following can give us much more bang for the buck:
+|  AngularJS v1.4-   |  AngularJS v1.5+   |  Angular v2.0+    |
+|:-----------------:|:-----------------:|:-----------------:|
+|  Directive        |Component Directive|      Component    |
+| are markers on a DOM element (i.e. attribute, element name, or CSS class) that tell AngularJS's HTML compiler to manipulate the DOM element and its children.  | are directives that define their own templates, controllers, and input/output bindings - the same things that Angular components define. | the main primitive from which user interfaces are built. You define the different portions of the ...
```
It might make sense to transpose this table 😛 (i.e. have the Angular versions in a column instead of a row).

gkalpak replied:
BTW, "Directive"/"Component Directive" (on the second row) are singular, while "are markers"/"are directives" (on the third row) are plural.

mary-poppins commented:
You can preview 0aaee40 at https://pr24692-0aaee40.ngbuilds.io/.
You can preview 0c8a860 at https://pr24692-0c8a860.ngbuilds.io/.
You can preview f8217bf at https://pr24692-f8217bf.ngbuilds.io/.
You can preview 9849a48 at https://pr24692-9849a48.ngbuilds.io/.
You can preview b0915a0 at https://pr24692-b0915a0.ngbuilds.io/.

kapunahelewong commented:
Closing due to lack of activity. If you'd still like to contribute, I'd be happy to review this change in a new PR.
